README.md
# Decimal.Khafra
An implementation of Decimal.js that uses native BigInts.

# Trade-offs
* Floats are converted to integers as BigInt doesn't support floats.
* Old(er) browsers don't support BigInts.
* No trig, alias, or "useless" functions (such as ``toBinary``). Functions that also only work with ``numbers`` won't be included.

# Benefits
* Value can never be NaN or non-BigInt (factory method will throw an error if such a value is passed).
* Should be relatively faster since it uses native JS methods instead of modifying strings.

CHANGELOG.md
# v1.0.0
* Initial release

# v1.0.1
* Add repo link in `package.json`.

# v1.0.2
* Add `<Decimal>.toObject`, `Decimal.pow`, and `<Decimal>.pow` functions.
* Allow strings like `1e+40` to be parsed.
* Fix a bad `<Decimal>.negate` test.

test/Decimal.test.js
const assert = require('assert');
const { Decimal } = require('../Decimal');

describe('Decimals can be created from strings, numbers, BigInts, or already existing Decimals.', () => {
    it('works with scientific notation strings and numbers', () => {
        assert.strictEqual(new Decimal('10e10').value, BigInt(10e10));
        assert.strictEqual(new Decimal('1.5e6').value, BigInt('1500000'));
        assert.strictEqual(new Decimal(1e6).value, 1000000n);
    });

    it('works with regular number-like strings and numbers', () => {
        assert.strictEqual(new Decimal('1000000').value, BigInt(1e6));
        assert.strictEqual(new Decimal('1'.padEnd(1000, '0')).value, BigInt('1'.padEnd(1000, '0')));
        assert.strictEqual(new Decimal(Number.MAX_SAFE_INTEGER).value, 9007199254740991n);
    });

    it('converts floats to integers in strings and numbers', () => {
        assert.strictEqual(new Decimal('43241386.5347951').value, 43241386n);
        assert.strictEqual(new Decimal(100000.005053920).value, 100000n);
    });

    it('throws when Infinity or NaN is passed through', () => {
        assert.throws(() => new Decimal(Infinity));
        assert.throws(() => new Decimal(1e1000));
        assert.throws(() => new Decimal(NaN));
    });

    it('Decimal#greaterThanOrEqualTo works as expected', () => {
        assert.strictEqual(new Decimal(1e6).greaterThanOrEqualTo(1e6), true);
        assert.strictEqual(new Decimal('1e6').greaterThanOrEqualTo(1e5), true);
        assert.strictEqual(new Decimal(1000000).greaterThanOrEqualTo(1e7), false);
    });

    it('Decimal#lessThanOrEqualTo works as expected', () => {
        assert.strictEqual(new Decimal(1e6).lessThanOrEqualTo(1e6), true);
        assert.strictEqual(new Decimal('1e6').lessThanOrEqualTo(1e5), false);
        assert.strictEqual(new Decimal(1000000).lessThanOrEqualTo(1e7), true);
    });

    it('Decimal#greaterThan works as expected', () => {
        assert.strictEqual(new Decimal(1e6).greaterThan(1e6), false);
        assert.strictEqual(new Decimal('1e6').greaterThan(1e5), true);
        assert.strictEqual(new Decimal(1000000).greaterThan(1e7), false);
    });

    it('Decimal#lessThan works as expected', () => {
        assert.strictEqual(new Decimal(1e6).lessThan(1e6), false);
        assert.strictEqual(new Decimal('1e6').lessThan(1e5), false);
        assert.strictEqual(new Decimal(1000000).lessThan(1e7), true);
    });

    it('Decimal#compareTo works as expected', () => {
        assert.strictEqual(new Decimal('1e100').compareTo('1e99'), 1);
        assert.strictEqual(new Decimal('1e100').compareTo('1e101'), -1);
        assert.strictEqual(new Decimal('1e100').compareTo('1e100'), 0);
        assert.throws(() => new Decimal('1e100').compareTo(NaN));
        assert.throws(() => new Decimal('1e100').compareTo(Infinity));
    });

    it('Decimal#divideBy works as expected', () => {
        assert.strictEqual(new Decimal('1'.padEnd(100, '0')).dividedBy('1000000') === BigInt('1'.padEnd(94, '0')), true);
        assert.strictEqual(new Decimal('10000').dividedBy(1000) === 10n, true);
        assert.strictEqual(new Decimal(new Decimal(10000n)).dividedBy(new Decimal(1000n)) === 10n, true);
    });

    it('Decimal#equals works as expected', () => {
        assert.strictEqual(new Decimal('10e1000').equals(new Decimal('10e1000')), true);
        assert.strictEqual(new Decimal('2'.padEnd(7, '0')).equals(2000000), true);
        assert.strictEqual(new Decimal(100000000000000000000000000000n).equals(1e7), false);
    });

    it('Decimal#toString works exactly as BigInt.toString does', () => {
        assert.strictEqual(new Decimal(1e6).toString(), '1000000');
        assert.strictEqual(new Decimal(1e6).toString(2), '11110100001001000000'); // binary
    });

    it('Decimal#add works as expected', () => {
        assert.strictEqual(new Decimal('10000').add(6000).equals(16000), true);
        assert.strictEqual(new Decimal('100000').add('-50000').value, BigInt(50000));
        assert.strictEqual(new Decimal('5050').add(50).add('900').add(100n).value, BigInt(6100));
    });

    it('Decimal#minus works as expected', () => {
        assert.strictEqual(new Decimal('10000').minus(6000).equals(4000n), true);
        assert.strictEqual(new Decimal('100000').minus('-50000').value, BigInt(150000)); 
        assert.strictEqual(new Decimal(5050n).minus(50).minus(900).minus(new Decimal(100)).equals(4000), true);
    });

    it('Decimal#multiply works as expected', () => {
        assert.strictEqual(new Decimal('10000').multiply(6000).equals(6e7), true);
        assert.strictEqual(new Decimal('100000').multiply('-50000').value, BigInt('-5000000000')); 
        assert.strictEqual(new Decimal(5050n).multiply(50).multiply(900).multiply(new Decimal(100)).equals(22725000000), true);
    });

    it('Decimal#modulo works as expected', () => {
        assert.strictEqual(new Decimal('1'.padEnd(1000, '0')).modulo(10), 0n);
        assert.strictEqual(new Decimal('1'.padEnd(1000, '0')).modulo(9), 1n);
    });

    it('Decimal#negate works as expected', () => {
        assert.strictEqual(new Decimal('1'.padEnd(1000, '0')).negate() === BigInt('-1'.padEnd(1001, '0')), true);
        assert.strictEqual(new Decimal('-1').negate() === BigInt(1), true);
    });

    it('Decimal#root works as expected', () => {
        assert.strictEqual(new Decimal(144).root() === 12n, true);
        assert.strictEqual(new Decimal(169).root() === 13n, true);
        assert.strictEqual(new Decimal(27).root(3) === 3n, true);
    });

    it('Decimal#pow works as expected', () => {
        assert.strictEqual(new Decimal('2').pow(3) === 8n, true);
        assert.strictEqual(new Decimal('4').pow(3) === 64n, true);
        assert.throws(() => new Decimal('4').pow(-2));
    });

    it('static Decimal#pow works as expected', () => {
        assert.strictEqual(Decimal.pow(2, 3) === 8n, true);
        assert.throws(() => Decimal.pow(2, -2))
    });
});

Decimal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Decimal = void 0;
/**
 * An implementation of Decimal.js that uses BigInts.
 * @author Khafra (KhafraDev)
 */
class Decimal {
    constructor(value) {
        this.value = this.factory(value);
    }
    /**
     * Create a Decimal object.
     */
    factory(value) {
        if (typeof value === 'bigint') {
            return value;
        }
        else if (typeof value === 'number') {
            if (value !== value) { // easy NaN check
                throw new Error('Factory method encountered NaN!');
            }
            else if (!Number.isFinite(value)) { // Infinity check
                throw new Error('Factory method encountered Infinity!');
            }
            return BigInt(Math.floor(value));
        }
        else if (typeof value === 'string') {
            if (+value === +value && isFinite(+value) && Number.isSafeInteger(+value)) { // string can be converted to number
                return BigInt(Math.floor(+value));
            }
            else if (value.indexOf('e') === -1) { // can be converted to BigInt easily
                return BigInt(value.split('.').shift());
            }
            // 1.4e1000 -> true
            // 1e1000 	-> true
            if (!/\d+e\d+/.test(value)) {
                throw new Error(`Invalid string received: ${value}.`);
            }
            else if (value.indexOf('.') > -1 && value.indexOf('e') === -1) {
                // 1.44020293 -> true
                // 1.4e100000 -> false
                return BigInt(value.split('.').shift());
            }
            const [mantissa, exponent] = value.split('e');
            const places = (+mantissa % 1 === 0) // 1.4 % 1 !== 0, 1 % 1 === 0
                ? +exponent
                : Math.abs((mantissa.length - 1) - +exponent);
            const intStr = mantissa.replace(/[^0-9\-]/, '').padEnd(places, '0');
            return BigInt(intStr);
        }
        else if (value instanceof Decimal) {
            return BigInt(value.value);
        }
        else {
            const toString = {}.toString;
            throw new Error(`Unexpected type in factory method: ${toString.call(value)}`);
        }
    }
    /**
     * Check if the current Decimal value is greater than or equal to another Decimal
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    greaterThanOrEqualTo(value) {
        return this.value >= this.factory(value);
    }
    /**
     * Check if the current Decimal value is less than or equal to another Decimal
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    lessThanOrEqualTo(value) {
        return this.value <= this.factory(value);
    }
    /**
     * Check if the current Decimal value is greater than another Decimal
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    greaterThan(value) {
        return this.value > this.factory(value);
    }
    /**
     * Check if the current Decimal value is less than another Decimal
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    lessThan(value) {
        return this.value < this.factory(value);
    }
    /**
     * @return
     * 	1 		if the value of the current Decimal is greater than the value of the input
     *
     * 	-1 		if the value of the current Decimal is less than the value of the input
     *
     * 	0		if they are equal
     * @param value Value to compare to
     */
    compareTo(value) {
        value = this.factory(value);
        if (Number.isNaN(value) || Number.isNaN(this.value)) {
            return NaN;
        }
        else if (this.value === value) {
            return 0;
        }
        else if (this.value > value) {
            return 1;
        }
        else { // this.value < value
            return -1;
        }
    }
    /**
     * Divide the current Decimal by a Decimal-like value.
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    dividedBy(value) {
        return this.value / this.factory(value);
    }
    /**
     * Returns `true` if values are equal, false otherwise.
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    equals(value) {
        return this.value === this.factory(value);
    }
    /**
     * @see {bigint.toString}
     * @param radix number
     */
    toString(radix) {
        return this.value.toString(radix);
    }
    /**
     * Subtract 2 Decimals.
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    minus(value) {
        return this.value = this.value - this.factory(value), this;
    }
    /**
     * Add 2 Decimals.
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    add(value) {
        return this.value = this.value + this.factory(value), this;
    }
    /**
     * Multiply 2 Decimals.
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    multiply(value) {
        return this.value = this.value * this.factory(value), this;
    }
    /**
     * Get remainder of a Decimal given a Decimal-like value
     * @param {DecimalInputs} value string | bigint | number | Decimal
     */
    modulo(value) {
        return this.value % this.factory(value);
    }
    /**
     * Return a negated Decimal.
     */
    negate() {
        return this.value * -1n;
    }
    /**
     * nth root
     * @see https://stackoverflow.com/a/58863398
     */
    root(k = 2n) {
        k = this.factory(k);
        let o = 0n;
        let x = this.value;
        let limit = 100n;
        while (x ** k !== k && x !== o && --limit) {
            o = x;
            x = ((k - 1n) * x + this.value / x ** (k - 1n)) / k;
        }
        return x;
    }
}
exports.Decimal = Decimal;


Decimal.ts
type DecimalInputs = number | bigint | Decimal | string;

/**
 * An implementation of Decimal.js that uses BigInts.
 * @author Khafra (KhafraDev)
 */
export class Decimal {
    value: bigint;
    constructor(value: DecimalInputs) {
        this.value = this.factory(value);
	}

	/**
	 * Create a Decimal object.
	 */
	factory(value: DecimalInputs) {
		if(typeof value === 'bigint') {
			return value;
		} else if(typeof value === 'number') {
			if(value !== value) { // easy NaN check
				throw new Error('Factory method encountered NaN!');
			} else if(!Number.isFinite(value)) { // Infinity check
				throw new Error('Factory method encountered Infinity!');
			}

			return BigInt(Math.floor(value));
		} else if(typeof value === 'string') {
			if(+value === +value && isFinite(+value) && Number.isSafeInteger(+value)) { // string can be converted to number
				return BigInt(Math.floor(+value));
			} else if(value.indexOf('e') === -1) { // can be converted to BigInt easily
				return BigInt(value.split('.').shift());
			}

			// 1.4e1000 -> true
			// 1e1000 	-> true
			if(!/\d+e\+?\d+/.test(value)) {
				throw new Error(`Invalid string received: ${value}.`);
			} else if(value.indexOf('.') > -1 && value.indexOf('e') === -1) { 
				// 1.44020293 -> true
				// 1.4e100000 -> false
				return BigInt(value.split('.').shift());
			}
			
			const [mantissa, exponent] = value.split(/e\+?/);
			const places = (+mantissa % 1 === 0) // 1.4 % 1 !== 0, 1 % 1 === 0
				? +exponent
				: Math.abs((mantissa.length - 1) - +exponent);
			const intStr = mantissa.replace(/[^0-9\-]/, '').padEnd(places, '0');

			return BigInt(intStr);
		} else if(value instanceof Decimal) { 
			return BigInt(value.value);
		} else {
			const toString = {}.toString;
			throw new Error(`Unexpected type in factory method: ${toString.call(value)}`)
		}
	}
	
	/**
	 * Check if the current Decimal value is greater than or equal to another Decimal
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	greaterThanOrEqualTo(value: DecimalInputs) {
		return this.value >= this.factory(value);
	}

	/**
	 * Check if the current Decimal value is less than or equal to another Decimal
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	lessThanOrEqualTo(value: DecimalInputs) {
		return this.value <= this.factory(value);
	}

	/**
	 * Check if the current Decimal value is greater than another Decimal
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	greaterThan(value: DecimalInputs) {
		return this.value > this.factory(value);
	}

	/**
	 * Check if the current Decimal value is less than another Decimal
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	lessThan(value: DecimalInputs) {
		return this.value < this.factory(value);
	}

	/**
	 * @return 
	 * 	1 		if the value of the current Decimal is greater than the value of the input
	 * 
	 * 	-1 		if the value of the current Decimal is less than the value of the input
	 * 
	 * 	0		if they are equal
	 * @param value Value to compare to
	 */
	compareTo(value: DecimalInputs) {
		value = this.factory(value);
		if(Number.isNaN(value) || Number.isNaN(this.value)) {
			return NaN;
		} else if(this.value === value) {
			return 0;
		} else if(this.value > value) {
			return 1;
		} else { // this.value < value
			return -1;
		}
	}

	/**
	 * Divide the current Decimal by a Decimal-like value.
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	dividedBy(value: DecimalInputs) {
		return this.value / this.factory(value);
	}

	/**
	 * Returns `true` if values are equal, false otherwise.
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	equals(value: DecimalInputs) {
		return this.value === this.factory(value);
	}

	/**
	 * @see {bigint.toString}
	 * @param radix number
	 */
    toString(radix?: number) {
        return this.value.toString(radix);
	}
	
	/**
	 * Subtract 2 Decimals.
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	minus(value: DecimalInputs) {
		return this.value = this.value - this.factory(value), this;
	}

	/**
	 * Add 2 Decimals.
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	add(value: DecimalInputs) {
		return this.value = this.value + this.factory(value), this;
	}

	/**
	 * Multiply 2 Decimals.
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	multiply(value: DecimalInputs) {
		return this.value = this.value * this.factory(value), this;
	}

	/**
	 * Get remainder of a Decimal given a Decimal-like value
	 * @param {DecimalInputs} value string | bigint | number | Decimal
	 */
	modulo(value: DecimalInputs) {
		return this.value % this.factory(value);
	}

	/**
	 * Return a negated Decimal.
	 */
	negate() {
		return this.value * -1n;
	}

	/**
	 * nth root
	 * @see https://stackoverflow.com/a/58863398
	 */
	root(k = 2n) {
		k = this.factory(k);
		let o = 0n;
		let x = this.value;
		let limit = 100n;
		
		while(x ** k !== k && x !== o && --limit) {
		  o = x;
		  x = ((k - 1n) * x + this.value / x ** (k - 1n)) / k;
		}
		
		return x;
	}

	/**
	 * multiply a Decimal by itself <value> times.
	 * @throws if value < 0n
	 */
	pow(value: DecimalInputs) {
		value = this.factory(value);
		if(value < 0n) {
			throw new Error(`Received negative power ${value} on <Decimal>.pow!`);
		}

        return this.value ** value;
    }

	/**
	 * multiply a given Decimal-like value by itself <value> times.
	 * is called on the Decimal class itself; not in an instance of Decimal.
	 * @throws if value < 0n
	 */
    static pow(a: DecimalInputs, b: DecimalInputs) {
        a = new Decimal(a);
        b = new Decimal(b);

		if(b.value < 0n) {
			throw new Error(`Received negative power ${b} on static Decimal.pow!`);
		}
		
		return a.value ** b.value;
    }

	/**
	 * Returns an Object representation of a Decimal class.
	 * Can be used in JSON.stringify to convert to and from BigInt.
	 */
    toObject() {
        return {
            value: this.value.toString()
        }
    }
}